sqlx-postgres/src/types/cube.rs

```diff
@@ -0,0 +1,251 @@
+use crate::decode::Decode;
+use crate::encode::{Encode, IsNull};
+use crate::types::Type;
+use crate::{PgArgumentBuffer, PgTypeInfo, PgValueRef, Postgres};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum PgCube {
+    Point(f64),
+    ZeroVolume(Vec<f64>),
+    OneDimensionInterval(f64, f64),
+    MultiDimension(Vec<Vec<f64>>),
+}
+
+impl Type<Postgres> for PgCube {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("cube")
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for PgCube {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        let bytes = value.as_bytes()?;
```

Comment:
This doesn't just need to be able to handle the binary format, but the text format as well (which would be used when querying using `sqlx::raw_sql()`).

Reply:
This is what the `test_unprepared_type_*` tests cover, by the way, which is likely why it's failing.

Reply:
Thanks! Really helpful.
Can see value.as_str() returning the text format similar to what you get in a Postgres tui/gui.
Will correct with a string parser 👍

Reply:
You want to match on [`PgValueRef::format()`](https://docs.rs/sqlx/latest/sqlx/postgres/struct.PgValueRef.html#method.format) to know which is which.

`value.as_str()` doesn't specifically care about text format vs binary format, it just runs UTF-8 validation on the raw bytes of the value.